File: python/model/time.py
from typing import List, Dict

class Time:
    def __init__(self, nome: str):
        self.nome = nome
        self.pontos = 0
        self.jogos = 0
        self.vitorias = 0
        self.empates = 0
        self.derrotas = 0
        self.gols_feitos = 0
        self.gols_sofridos = 0
        self.historico_rodadas = []
        self.posicoes_por_rodada = {}  # rodada -> posição
    
    def reiniciar(self):
        """Reinicia todas as estatísticas do time"""
        self.pontos = 0
        self.jogos = 0
        self.vitorias = 0
        self.empates = 0
        self.derrotas = 0
        self.gols_feitos = 0
        self.gols_sofridos = 0
        self.historico_rodadas.clear()
        self.posicoes_por_rodada.clear()
    
    def adicionar_resultado_sem_posicao(self, gols_feitos: int, gols_sofridos: int, rodada: int):
        """Adiciona resultado de uma partida sem calcular posição"""
        self.jogos += 1
        self.gols_feitos += gols_feitos
        self.gols_sofridos += gols_sofridos
        
        # Adicionar rodada se ainda não foi adicionada
        if rodada not in self.historico_rodadas:
            self.historico_rodadas.append(rodada)
        
        # Calcular pontos
        if gols_feitos > gols_sofridos:
            self.vitorias += 1
            self.pontos += 3
        elif gols_feitos == gols_sofridos:
            self.empates += 1
            self.pontos += 1
        else:
            self.derrotas += 1
    
    def atualizar_posicao_para_rodada(self, rodada: int, posicao: int):
        """Atualiza a posição para uma rodada específica"""
        self.posicoes_por_rodada[rodada] = posicao
    
    def get_saldo_gols(self) -> int:
        """Retorna o saldo de gols"""
        return self.gols_feitos - self.gols_sofridos
    
    def get_aproveitamento(self) -> float:
        """Retorna o aproveitamento em percentual"""
        if self.jogos == 0:
            return 0.0
        return (self.pontos * 100.0) / (self.jogos * 3)
    
    def get_historico_rodadas(self) -> List[int]:
        """Retorna as rodadas ordenadas"""
        return sorted(self.historico_rodadas)
    
    def get_historico_posicoes(self) -> List[int]:
        """Retorna as posições ordenadas por rodada"""
        rodadas_ordenadas = self.get_historico_rodadas()
        posicoes_ordenadas = []
        
        for rodada in rodadas_ordenadas:
            posicao = self.posicoes_por_rodada.get(rodada, 20)  # Default 20
            posicoes_ordenadas.append(posicao)
        
        return posicoes_ordenadas
    
    def imprimir_historico(self):
        """Imprime o histórico para debug"""
        print(f"=== Histórico do {self.nome} ===")
        rodadas = self.get_historico_rodadas()
        posicoes = self.get_historico_posicoes()
        
        for rodada, posicao in zip(rodadas, posicoes):
            print(f"Rodada {rodada}: {posicao}ª posição")
        print("============================")

File: java/src/main/java/model/Time.java
package model;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Time {
    private String nome;
    private int pontos;
    private int jogos;
    private int vitorias;
    private int empates;
    private int derrotas;
    private int golsFeitos;
    private int golsSofridos;
    private List<Integer> historicoRodadas;
    private Map<Integer, Integer> posicoesPorRodada; // NOVA estrutura: rodada -> posição
    
    public Time(String nome) {
        this.nome = nome;
        this.pontos = 0;
        this.jogos = 0;
        this.vitorias = 0;
        this.empates = 0;
        this.derrotas = 0;
        this.golsFeitos = 0;
        this.golsSofridos = 0;
        this.historicoRodadas = new ArrayList<>();
        this.posicoesPorRodada = new HashMap<>();
    }
    
    public void reiniciar() {
        this.pontos = 0;
        this.jogos = 0;
        this.vitorias = 0;
        this.empates = 0;
        this.derrotas = 0;
        this.golsFeitos = 0;
        this.golsSofridos = 0;
        this.historicoRodadas.clear();
        this.posicoesPorRodada.clear();
    }
    
    // MÉTODO CORRIGIDO: Adiciona resultado sem posição
    public void adicionarResultadoSemPosicao(int golsFeitos, int golsSofridos, int rodada) {
        this.jogos++;
        this.golsFeitos += golsFeitos;
        this.golsSofridos += golsSofridos;
        
        // Adicionar rodada se ainda não foi adicionada
        if (!historicoRodadas.contains(rodada)) {
            this.historicoRodadas.add(rodada);
        }
        
        if (golsFeitos > golsSofridos) {
            this.vitorias++;
            this.pontos += 3;
        } else if (golsFeitos == golsSofridos) {
            this.empates++;
            this.pontos += 1;
        } else {
            this.derrotas++;
        }
    }
    
    // NOVO MÉTODO: Atualiza posição para uma rodada específica
    public void atualizarPosicaoParaRodada(int rodada, int posicao) {
        posicoesPorRodada.put(rodada, posicao);
    }
    
    // MÉTODO CORRIGIDO: Retorna posições ordenadas por rodada
    public List<Integer> getHistoricoPosicoes() {
        List<Integer> rodadasOrdenadas = new ArrayList<>(historicoRodadas);
        rodadasOrdenadas.sort(Integer::compareTo);
        
        List<Integer> posicoesOrdenadas = new ArrayList<>();
        for (Integer rodada : rodadasOrdenadas) {
            Integer posicao = posicoesPorRodada.get(rodada);
            posicoesOrdenadas.add(posicao != null ? posicao : 20); // Default 20 se não encontrar
        }
        
        return posicoesOrdenadas;
    }
    
    // Método antigo mantido para compatibilidade (mas não usado)
    @Deprecated
    public void adicionarResultado(int golsFeitos, int golsSofridos, int rodada, int posicao) {
        adicionarResultadoSemPosicao(golsFeitos, golsSofridos, rodada);
        atualizarPosicaoParaRodada(rodada, posicao);
    }
    
    public int getSaldoGols() {
        return golsFeitos - golsSofridos;
    }
    
    public double getAproveitamento() {
        if (jogos == 0) return 0.0;
        return (pontos * 100.0) / (jogos * 3);
    }
    
    // Getters
    public String getNome() { return nome; }
    public int getPontos() { return pontos; }
    public int getJogos() { return jogos; }
    public int getVitorias() { return vitorias; }
    public int getEmpates() { return empates; }
    public int getDerrotas() { return derrotas; }
    public int getGolsFeitos() { return golsFeitos; }
    public int getGolsSofridos() { return golsSofridos; }
    
    public List<Integer> getHistoricoRodadas() { 
        List<Integer> rodadasOrdenadas = new ArrayList<>(historicoRodadas);
        rodadasOrdenadas.sort(Integer::compareTo);
        return rodadasOrdenadas;
    }
    
    // NOVO MÉTODO: Para debug
    public void imprimirHistorico() {
        System.out.println("=== Histórico do " + nome + " ===");
        List<Integer> rodadas = getHistoricoRodadas();
        List<Integer> posicoes = getHistoricoPosicoes();
        
        for (int i = 0; i < rodadas.size(); i++) {
            System.out.printf("Rodada %d: %dª posição%n", rodadas.get(i), posicoes.get(i));
        }
        System.out.println("============================");
    }
}